File: 1/final.ml
let rec read file =
  try 
      let line = input_line file in
      line :: (read file)
  with End_of_file -> []

let rec lines_to_lists lines l r =
  match lines with
  | [] -> (List.rev l, List.rev r)
  | line :: rest ->
      let (a, b) = Scanf.sscanf line "%d %d" (fun a b -> (a, b)) in
      lines_to_lists rest (a :: l) (b :: r)
      
let rec counter l r =
  match l with
  | [] -> 0
  | h :: t -> if h = r then 1 + counter t r else counter t r

let rec part1 a b =
  match (a, b) with
  | [], [] -> 0
  | a :: rest_a, b :: rest_b ->
      abs (a - b) + part1 rest_a rest_b
  | _ -> failwith "Invalid input"

let rec part2 a b =
  match a with
  | [] -> 0
  | h :: t -> h*(counter b h) + part2 t b

let () =
  let lines = read (open_in "1/input.txt") in
  let (a, b) = lines_to_lists lines [] [] in
  let c = List.sort compare a in
  let d = List.sort compare b in
  Printf.printf "part1 : %d\n" (part1 c d);
  Printf.printf "part2 : %d\n" (part2 c d)

File: 1/final.py
def read_file(file_name):
    with open(file_name, 'r') as file:
        l = [tuple(map(int, line.strip().split())) for line in file]
    return [list(x) for x in zip(*l)]
            
def part1(l, r):
    l.sort()
    r.sort()
    return sum(abs(a-b) for a, b in zip(l, r))

def part2(l, r):
    from collections import Counter
    s = 0
    c = Counter(l)
    d = Counter(r)
    return sum(k*c[k]*d[k] for k in c)

if __name__ == "__main__":
    a, b = read_file("1/input.txt")
    print("part1:", part1(a, b))
    print("part2:", part2(a, b))
    